### PR TITLE
Add `EncoderLayer`/`DecoderLayer`

### PR DESCRIPTION
**Pull request by danieldk**

### CODE REVIEW

File: src/layers/transformer.rs [hunk 1]
```diff
@@ -154,24 +158,376 @@ impl TransformerEmbeddings {
...
+    ///   layers.
+    /// * `feed_forward_layer` - The feed-forward layer to use in the
+    ///   transformer layers.
+    /// * `dropouts` - The dropouts to use in the transformer layers.
+    /// * `layer_norms` - The layer norms to use in the transformer
+    /// * `use_parallel_attention` - Use parallel attention.
+    fn new(
+        attention_layer: SelfAttention,
+        feed_forward_layer: PointwiseFeedForward,
+        dropouts: TransformerDropouts,
```
Comment (shadeMe) [inline src/layers/transformer.rs]:
```suggestion
        dropouts: TransformerDropouts,
        feed_forward_layer: PointwiseFeedForward,
```

File: src/layers/transformer.rs [hunk 2]
```diff
@@ -154,24 +158,376 @@ impl TransformerEmbeddings {
...
+            use_parallel_attention,
+        }
+    }
+
+    /// Apply the transformer layer to the given piece hidden representations.
+    ///
+    /// * `input` - Hidden representations to apply the layer to.
+    ///   *Shape:* `(batch_size, seq_len, width)`
+    /// * `attention_mask` - Attention mask. Sequence elements for which the
+    ///    corresponding mask element is set to `False` are ignored
```
Comment (shadeMe) [inline src/layers/transformer.rs]:
```suggestion
    ///    corresponding mask element is set to `false` are ignored
```

File: src/layers/transformer.rs [hunk 3]
```diff
@@ -154,24 +158,376 @@ impl TransformerEmbeddings {
...
+    ///    during attention calculation.
+    /// * `use_causal_mask` - Mask out succeeding sequence elements when `True`.
```
Comment (shadeMe) [inline src/layers/transformer.rs]:
```suggestion
    /// * `use_causal_mask` - Mask out succeeding sequence elements when `true`.
```

File: src/layers/transformer.rs [hunk 4]
```diff
@@ -154,24 +158,376 @@ impl TransformerEmbeddings {
...
+    /// *Shape:* ``(batch_size, seq_len, width)``
+    #[allow(clippy::too_many_arguments)]
+    fn forward(
+        &self,
+        input: &Tensor,
+        train: bool,
+        attention_mask: &AttentionMask,
+        _cache: Option<&KeyValueCache>,
+        _positions: Option<&Tensor>,
+        _store_cache: bool,
```
Comment (shadeMe) [inline src/layers/transformer.rs]:
```suggestion
        attention_mask: &AttentionMask,
        _cache: Option<&KeyValueCache>,
        _positions: Option<&Tensor>,
        _store_cache: bool,
        train: bool, 
```
Unless we want to make `train` special (since it's also exposed by `ModuleT`) and keep it as the second argument?

File: src/layers/transformer.rs [hunk 5]
```diff
@@ -154,24 +158,376 @@ impl TransformerEmbeddings {
...
+    /// * `cache` - Key/value cache to avoid recomputing key/value representations
+    ///    for tokens that were previously seen.
+    /// * `positions` - Input positions. Positions are needed to look up rotary
+    ///    embeddings. Normally, these positions are calculated automatically.
+    ///    But if the positions deviate for some reason, they can be provided
+    ///    through this argument.
+    ///    *Shape:* `(batch_size, seq_len)`
+    /// * `store_cache` - Whether to cache the key/value representations for
+    ///   future reuse.
+    ///
```
Comment (shadeMe) [inline src/layers/transformer.rs:324]: Missing docstring for `train`.

File: src/layers/transformer.rs [hunk 6]
```diff
@@ -154,24 +158,376 @@ impl TransformerEmbeddings {
...
+}
+
+impl DecoderLayer {
+    /// Construct a decoder layer.
+    ///
+    /// * `attention_layer` - The attention layer to use in the transformer
+    ///   layers.
+    /// * `feed_forward_layer` - The feed-forward layer to use in the
+    ///   transformer layers.
+    /// * `dropouts` - The dropouts to use in the transformer layers.
```
Comment (shadeMe) [inline src/layers/transformer.rs]:
```suggestion
    /// * `dropouts` - The dropouts to use in the transformer layers.
    /// * `feed_forward_layer` - The feed-forward layer to use in the
    ///   transformer layers.
```

File: src/layers/transformer.rs [hunk 7]
```diff
@@ -154,24 +158,376 @@ impl TransformerEmbeddings {
...
+    /// * `layer_norms` - The layer norms to use in the transformer
+    /// * `use_parallel_attention` - Use parallel attention.
+    pub fn new(
+        attention_layer: SelfAttention,
+        feed_forward_layer: PointwiseFeedForward,
+        dropouts: TransformerDropouts,
```
Comment (shadeMe) [inline src/layers/transformer.rs]:
```suggestion
        dropouts: TransformerDropouts,
        feed_forward_layer: PointwiseFeedForward,
```

File: src/layers/transformer.rs [hunk 8]
```diff
@@ -154,24 +158,376 @@ impl TransformerEmbeddings {
...
+    /// * `cache` - Key/value cache to avoid recomputing key/value representations
+    ///    for tokens that were previously seen.
+    /// * `positions` - Input positions. Positions are needed to look up rotary
+    ///    embeddings. Normally, these positions are calculated automatically.
+    ///    But if the positions deviate for some reason, they can be provided
+    ///    through this argument.
+    ///    *Shape:* `(batch_size, seq_len)`
+    /// * `store_cache` - Whether to cache the key/value representations for
+    ///   future reuse.
+    ///
```
Comment (shadeMe) [inline src/layers/transformer.rs:438]: Missing `train` docstring.

File: src/layers/transformer.rs [hunk 9]
```diff
@@ -154,24 +158,376 @@ impl TransformerEmbeddings {
...
+    /// Returns layer output and the key/value cache.
+    /// *Shape:* ``(batch_size, seq_len, width)``
+    pub fn forward(
+        &self,
+        input: &Tensor,
+        train: bool,
```
Comment (shadeMe) [inline src/layers/transformer.rs]: Same question as above.

File: src/layers/transformer.rs [hunk 10]
```diff
@@ -154,24 +158,376 @@ impl TransformerEmbeddings {
...
+}
+
+impl EncoderLayer {
+    /// Construct an encoder layer.
+    ///
+    /// * `attention_layer` - The attention layer to use in the transformer
+    ///   layers.
+    /// * `feed_forward_layer` - The feed-forward layer to use in the
+    ///   transformer layers.
+    /// * `dropouts` - The dropouts to use in the transformer layers.
```
Comment (shadeMe) [inline src/layers/transformer.rs]:
```suggestion
    /// * `dropouts` - The dropouts to use in the transformer layers.
    /// * `feed_forward_layer` - The feed-forward layer to use in the
    ///   transformer layers.
```

File: src/layers/transformer.rs [hunk 11]
```diff
@@ -154,24 +158,376 @@ impl TransformerEmbeddings {
...
+    /// * `layer_norms` - The layer norms to use in the transformer
+    /// * `use_parallel_attention` - Use parallel attention.
+    pub fn new(
+        attention_layer: SelfAttention,
+        feed_forward_layer: PointwiseFeedForward,
+        dropouts: TransformerDropouts,
```
Comment (shadeMe) [inline src/layers/transformer.rs]:
```suggestion
        dropouts: TransformerDropouts,
        feed_forward_layer: PointwiseFeedForward,
```

File: src/layers/transformer.rs [hunk 12]
```diff
@@ -154,24 +158,376 @@ impl TransformerEmbeddings {
...
+    ///    *Shape:* `(batch_size, seq_len)`
+    /// * `store_cache` - Whether to cache the key/value representations for
+    ///   future reuse.
+    ///
+    /// Returns layer output and the key/value cache.
+    /// *Shape:* ``(batch_size, seq_len, width)``
+    pub fn forward(
+        &self,
+        input: &Tensor,
+        train: bool,
```
Comment (shadeMe) [inline src/layers/transformer.rs]: Same question as above.

File: src/layers/transformer.rs [hunk 13]
```diff
@@ -154,24 +158,376 @@ impl TransformerEmbeddings {
...
+    /// * `cache` - Key/value cache to avoid recomputing key/value representations
+    ///    for tokens that were previously seen.
+    /// * `positions` - Input positions. Positions are needed to look up rotary
+    ///    embeddings. Normally, these positions are calculated automatically.
+    ///    But if the positions deviate for some reason, they can be provided
+    ///    through this argument.
+    ///    *Shape:* `(batch_size, seq_len)`
+    /// * `store_cache` - Whether to cache the key/value representations for
+    ///   future reuse.
+    ///
```
Comment (shadeMe) [inline src/layers/transformer.rs:514]: Missing `train` docstring.

**Review by shadeMe — APPROVED** [review]
Just some minor comments on style and a couple of missing docstrings. Good to merge afterwards.

File: src/layers/transformer.rs (hunk 4 above)
Comment (danieldk) [reply src/layers/transformer.rs]: Yeah, that was my motivation to move it to the second position. Not sure though, we need to come up with a sensible ordering, maybe something like:

* inputs/xs
* Other input arguments in alphabetical order (attention mask, positions, cache).
* Boolean/option arguments in alphabetical order.

The only thing I don't really like is that an argument which is pretty much mandatory (train) ends up shifting its position all the time. On the other hand, reserving a position for it will only make the ordering rules more complex.

Comment (shadeMe) [reply src/layers/transformer.rs]: The above order sounds good to me!
